community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RelationshipGroupCache.java

```diff
@@ -0,0 +1,269 @@
+        {
+            short count = groupCountCache.getShort( nodeId, 0 );
+            if ( highCacheId + count > cache.length() )
+            {
+                // Cannot include this one, so up until the previous is good
+                return this.toNodeId = nodeId;
+            }
+            offsets.set( rebase( nodeId ), highCacheId );
+            highCacheId += count;
+        }
+        return this.toNodeId = highNodeId;
+    }
+
+    private long rebase( long toNodeId )
+    {
+        return toNodeId - fromNodeId;
+    }
+
+    public boolean put( RelationshipGroupRecord groupRecord )
+    {
+        long nodeId = groupRecord.getOwningNode();
+        assert nodeId < highNodeId;
+        if ( nodeId < fromNodeId || nodeId >= toNodeId )
+        {
+            return false;
+        }
+
+        long baseIndex = offsets.get( rebase( nodeId ) );
+        // grouCount is extra validation, really
+        int groupCount = groupCountCache.getShort( nodeId, 0 );
+        long index = scanForFreeFrom( baseIndex, groupCount, groupRecord.getType() );
+
+        // Put the group at this index
+        cache.setByte( index, 0, (byte) 1 );
+        cache.setShort( index, 1, (short) groupRecord.getType() );
+        cache.set6ByteLong( index, 1 + 2, groupRecord.getFirstOut() );
+        cache.set6ByteLong( index, 1 + 2 + 6, groupRecord.getFirstIn() );
+        cache.set6ByteLong( index, 1 + 2 + 6 + 6, groupRecord.getFirstLoop() );
+        return true;
+    }
+
+    private long scanForFreeFrom( long startIndex, int groupCount, int type )
+    {
+        long desiredIndex = -1;
+        long freeIndex = -1;
+        for ( int i = 0; i < groupCount; i++ )
+        {
+            long candidateIndex = startIndex + i;
+            boolean free = cache.getByte( candidateIndex, 0 ) == 0;
+            if ( free )
+            {
+                freeIndex = candidateIndex;
+                break;
+            }
+
+            if ( desiredIndex == -1 )
+            {
+                int existingType = cache.getShort( candidateIndex, 1 ) & 0xFFFF;
+                if ( existingType == type )
+                {
+                    throw new IllegalStateException( "Tried to put multiple groups with same type " + type );
+                }
+
+                if ( type < existingType )
+                {
+                    // This means that the groups have arrived here out of order, please put this group
+                    // in the correct place, not at the end
+                    desiredIndex = candidateIndex;
+                }
+            }
+        }
+
+        if ( freeIndex == -1 )
+        {
+            throw new IllegalStateException( "I can't believe it, there's no room for me for startIndex:" + startIndex +
+                    " with a group count of " + groupCount );
+        }
+
+        // TODO: Instead of doing the sorting here right away be doing the relatively expensive move
```

Comment:
If we remove the TODO I think it's fine. The comment is for aiding in extending this code if we need a feature which I can foresee us needing, but don't actually need right now.